File: infra/protocol/go/codec/codec.go
package codec

import (
	"encoding/json"
	"errors"

	"google.golang.org/protobuf/proto"
)

type Codec[T any] interface {
	ContentType() string
	Encode(T) ([]byte, error)
	Decode([]byte) (T, error)
}

type RawBytesCodec struct{}

func (RawBytesCodec) ContentType() string {
	return "application/octet-stream"
}

func (RawBytesCodec) Encode(value []byte) ([]byte, error) {
	return append([]byte(nil), value...), nil
}

func (RawBytesCodec) Decode(data []byte) ([]byte, error) {
	return append([]byte(nil), data...), nil
}

type JSONCodec[T any] struct{}

func (JSONCodec[T]) ContentType() string {
	return "application/json"
}

func (JSONCodec[T]) Encode(value T) ([]byte, error) {
	return json.Marshal(value)
}

func (JSONCodec[T]) Decode(data []byte) (T, error) {
	var value T
	err := json.Unmarshal(data, &value)
	return value, err
}

type ProtobufCodec[T proto.Message] struct {
	New func() T
}

func (ProtobufCodec[T]) ContentType() string {
	return "application/protobuf"
}

func (ProtobufCodec[T]) Encode(value T) ([]byte, error) {
	return proto.Marshal(value)
}

func (codec ProtobufCodec[T]) Decode(data []byte) (T, error) {
	var zero T
	if codec.New == nil {
		return zero, errors.New("protobuf codec requires New")
	}
	value := codec.New()
	err := proto.Unmarshal(data, value)
	return value, err
}

File: infra/protocol/python/pacific_rim_protocol/codec.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Callable, Generic, Protocol, TypeVar

T = TypeVar("T")


class Codec(Protocol[T]):
  content_type: str

  def encode(self, value: T) -> bytes:
    ...

  def decode(self, data: bytes) -> T:
    ...


@dataclass(frozen=True)
class RawBytesCodec:
  content_type: str = "application/octet-stream"

  def encode(self, value: bytes | bytearray | memoryview) -> bytes:
    return bytes(value)

  def decode(self, data: bytes) -> bytes:
    return bytes(data)


@dataclass(frozen=True)
class JsonCodec:
  content_type: str = "application/json"

  def encode(self, value: Any) -> bytes:
    return json.dumps(value, ensure_ascii=False, separators=(",", ":")).encode("utf-8")

  def decode(self, data: bytes) -> Any:
    return json.loads(data.decode("utf-8"))


class ProtobufCodec(Generic[T]):
  content_type = "application/protobuf"

  def __init__(self, message_factory: Callable[[], T] | type[T]):
    self._message_factory = message_factory

  def encode(self, value: T) -> bytes:
    serializer = getattr(value, "SerializeToString", None)
    if serializer is None:
      raise TypeError("protobuf value must expose SerializeToString()")
    return bytes(serializer())

  def decode(self, data: bytes) -> T:
    message = self._message_factory()
    parser = getattr(message, "ParseFromString", None)
    if parser is None:
      raise TypeError("protobuf message must expose ParseFromString()")
    parser(data)
    return message
